readme.md
# Blood Sugar Monitor Module

This module contains a block for the placement on any page, so as to record blood sugar values,
and a page to view the logged in users submitted data.

## Install

Place the project directory in the modules folder for the site, activate it in the drupal modules list,
and place the block on a page or toolbar of your choosing.

## Credits

Spec proposed by [Ashfield](https://www.ashfieldhealthcare.com/).

Developed as a learning experience by [indiealexh](https://github.com/indiealexh).

## Licence

The MIT License (MIT)
Copyright (c) 2016 Alexander Haslam

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


changelog.md
# 0.1.0
- Updated readme and some documentation

# 0.0.7
- Added table with BSM list for user.
- Added time checking for submission.

# 0.0.6
- Removed unnecessary commented out code
- Added database installation schema
- Added ability to record BS values
- Added BS Value Validation

# 0.0.5
- Added form Ajax
- Added success Ajax response

# 0.0.4
- Added form
- Added number field to form with validation for range of 1.00 to 10.00
- Added form to the block
- Added missing permissions.yml
- Fixed default permissions in install file

# 0.0.3
- Added placeholder page to view stored bsm data
- Added default link in main manu that points to the placeholder page
- Added additional TBA permissions
- Fixed missing annotations for the bsm block

# 0.0.2
- Added missing composer.json to boilerplate
- Added Block class

# 0.0.1
- Added boilerplate


src/Plugin/Block/BloodSugarMonitorBlock.php
<?php
/**
 * Provides a 'Hello' Block
 *
 * @Block(
 *   id = "blood_sugar_monitor_block",
 *   admin_label = @Translation("Blood Sugar Monitor block"),
 * )
 */

namespace Drupal\blood_sugar_monitor\Plugin\Block;

use Drupal\Core\Access\AccessResult;
use Drupal\Core\Block\BlockBase;
use Drupal\Core\Session\AccountInterface;

class BloodSugarMonitorBlock extends BlockBase{

  /**
   * Builds and returns the renderable array for this block plugin.
   *
   * If a block should not be rendered because it has no content, then this
   * method must also ensure to return no content: it must then only return an
   * empty array, or an empty array with #cache set (with cacheability metadata
   * indicating the circumstances for it being empty).
   *
   * @return array
   *   A renderable array representing the content of the block.
   *
   * @see \Drupal\block\BlockViewBuilder
   */
  public function build() {

    $render = array();
    $form = \Drupal::formBuilder()->getForm('Drupal\blood_sugar_monitor\Form\BloodSugarMonitorForm');
    $render[] = $form;

    return $render;
  }

  /**
   * @param \Drupal\Core\Session\AccountInterface $account
   * @return bool
   */
  public function blockAccess(AccountInterface $account) {
    if($account->hasPermission('store bsm data')) return AccessResult::allowed();
    else return AccessResult::forbidden();
  }

}


src/Controller/BloodSugarMonitorController.php
<?php
/**
 * @file
 * Contains \Drupal\blood_sugar_monitor\BloodSugarMonitorController.
 */

namespace Drupal\blood_sugar_monitor\Controller;

use Drupal\Core\Controller\ControllerBase;
use Drupal\Core\Database\Database;

class BloodSugarMonitorController extends ControllerBase{

  public function content() {

    $userid = \Drupal::currentUser()->id();

    $listings = Database::getConnection()
      ->select('blood_sugar_monitor','bsm')
      ->fields('bsm')
      ->range(0,72)
      ->condition('bsm.uid',$userid)
      ->orderBy('bsm.created', 'DESC')
      ->execute();

    $rows = [];

    foreach($listings as $listing) {
      $rows[]['data'] = [
        $listing->value,
        date('Y-m-d H:i:s',$listing->created),
      ];
    }

    return array(
      '#theme' => 'table',
      '#header' => [t('Value'),t('Recorded')],
      '#rows' => $rows,
    );
  }

}


src/Form/BloodSugarMonitorForm.php
<?php
/**
 * @file
 * Contains \Drupal\blood_sugar_monitor\Form\BloodSugarMonitorForm.
 */

namespace Drupal\blood_sugar_monitor\Form;

use Drupal\Core\Ajax\AjaxResponse;
use Drupal\Core\Ajax\ChangedCommand;
use Drupal\Core\Ajax\CssCommand;
use Drupal\Core\Ajax\HtmlCommand;
use Drupal\Core\Ajax\InvokeCommand;
use Drupal\Core\Database\Connection;
use Drupal\Core\Form\FormBase;
use Drupal\Core\Form\FormStateInterface;
use Drupal\Core\Database\Database;

class BloodSugarMonitorForm extends FormBase{
  /**
   * Returns a unique string identifying the form.
   *
   * @return string
   *   The unique string identifying the form.
   */
  public function getFormId() {
    return 'bsm_form';
  }

  /**
   * Form constructor.
   *
   * @param array $form
   *   An associative array containing the structure of the form.
   * @param \Drupal\Core\Form\FormStateInterface $form_state
   *   The current state of the form.
   *
   * @return array
   *   The form structure.
   */
  public function buildForm(array $form, FormStateInterface $form_state) {
    $form['blood_sugar'] = array(
      '#type' => 'number',
      '#title' => t('blood sugar level'),
      '#description' => 'Please enter a blood sugar value',
      '#default_value' => 5.00,
      '#min' => 1.00,
      '#max' => 10.00,
      '#step' => 0.01,
      '#required' => TRUE,
      '#suffix' => t('<div id="search-suggest"></div>'),
    );
    $form['submit_blood_sugar'] = array(
      '#type' => 'submit',
      '#value' => t('Add blood sugar value'),
      '#ajax' => array(
        'callback' => array($this, 'ajaxCallback'),
        'event' => 'click',
        'progress' => array(
          'type' => 'throbber',
          'message' => 'Submitting',
        ),
      ),
    );

    return $form;
  }

  protected function checkHourPassed() {

    $userid = \Drupal::currentUser()->id();

    $count = Database::getConnection()
      ->select('blood_sugar_monitor','bsm')
      ->fields('bsm')
      ->range(0,1)
      ->condition('bsm.uid',$userid)
      ->orderBy('bsm.created', 'DESC')
      ->countQuery()->execute()->fetchField();

    if((int)$count == (int)0) return true;

    $listings = Database::getConnection()
      ->select('blood_sugar_monitor','bsm')
      ->fields('bsm')
      ->range(0,1)
      ->condition('bsm.uid',$userid)
      ->orderBy('bsm.created', 'DESC')
      ->execute();

    foreach($listings as $listing) {
      $tpHour = (float)$listing->created + (float)3600;
      if((float)time() > (float)$tpHour) {
        return true;
      }
    }

    return false;
  }

  protected function validateBSValue($value) {
    $value = (float)$value;
    $valid = true;
    if(!is_float($value)) $valid = false;
    if($value < 1.00 || $value > 10.00) $valid = false;
    if(!preg_match('/^\d{1,2}\.?\d?\d?$/',$value)) $valid = false;
    return $valid;
  }

  protected function successResponse() {
    $response = new AjaxResponse();
    $user = \Drupal\user\Entity\User::load(\Drupal::currentUser()->id());
    $response->addCommand(new HtmlCommand('#bsm-form',t('Thank you '.$user->get('name')->value.', please come back in an hour to submit your next blood sugar value')));
    $response->addCommand(new CssCommand('#bsm-form',[
      'background' => '#f3faef no-repeat 10px 17px',
      'padding' => '15px 20px 15px 35px',
      'border' => '1px solid #c9e1bd',
      'border-radius' => '2px',
      'box-shadow' => '-8px 0 0 #77b259',
      'background-image' => 'url(/core/misc/icons/73b355/check.svg)'
    ]));
    return $response;
  }

  protected function timeErrorResponse() {
    $response = new AjaxResponse();
    $user = \Drupal\user\Entity\User::load(\Drupal::currentUser()->id());
    $response->addCommand(new HtmlCommand('#bsm-form',t('Sorry '.$user->get('name')->value.', you already submitted a value within the past hour')));
    $response->addCommand(new CssCommand('#bsm-form',[
      'background' => '#fcf4f2 no-repeat 10px 17px',
      'padding' => '15px 20px 15px 35px',
      'border' => '1px solid #f9c9bf',
      'border-radius' => '2px',
      'box-shadow' => '-8px 0 0 #e62600',
      'background-image' => 'url(/core/misc/icons/e32700/error.svg)'
    ]));
    return $response;
  }

  public function ajaxCallback(array &$form, FormStateInterface $form_state) {
    $user = \Drupal\user\Entity\User::load(\Drupal::currentUser()->id());

    if(!$this->checkHourPassed()) {
      return $this->timeErrorResponse();
    }

    if(!$this->validateBSValue($form_state->getValue('blood_sugar'))) {
      throw new \Exception('Error');
    }

    try {
      $fields = array(
        'uid' => \Drupal::currentUser()->id(),
        'value' => $form_state->getValue('blood_sugar'),
        'created' => time(),
      );
      Database::getConnection()->insert('blood_sugar_monitor')->fields($fields)->execute();

      return $this->successResponse();
    } catch (\Exception $e) {
      $response = new AjaxResponse();
      $response->addCommand(new CssCommand('#edit-blood-sugar',['border-color' => 'red']));
      $response->addCommand(new CssCommand('#edit-blood-sugar--description',['color' => 'red']));
      $response->addCommand(new HtmlCommand('#edit-blood-sugar--description',t('Invalid')));
      return $response;
    }

  }

  /**
   * Form submission handler.
   *
   * @param array $form
   *   An associative array containing the structure of the form.
   * @param \Drupal\Core\Form\FormStateInterface $form_state
   *   The current state of the form.
   */
  public function submitForm(array &$form, FormStateInterface $form_state) {
    //ToDo: Decide what to do with this
  }
}
